src/Lykke.Job.CandleHistoryWriter.Repositories/HistoryMigration/HistoryProviders/TradesSQLHistory/TradesSqlHistoryRepository.cs

```diff
@@ -2,45 +2,63 @@
 using System.Collections.Generic;
 using System.Data;
 using System.Data.SqlClient;
+using System.Linq;
 using System.Text;
 using System.Threading.Tasks;
 using Common.Log;
 using Lykke.Job.CandlesHistoryWriter.Core.Domain.HistoryMigration.HistoryProviders.TradesSQLHistory;
+using Lykke.Job.CandlesProducer.Contract;
 
 namespace Lykke.Job.CandleHistoryWriter.Repositories.HistoryMigration.HistoryProviders.TradesSQLHistory
 {
     public class TradesSqlHistoryRepository : ITradesSqlHistoryRepository, IDisposable
     {
         private readonly int _sqlQueryBatchSize;
-        private readonly ILog _log;
-
         private readonly SqlConnection _sqlConnection;
+        private readonly TimeSpan _sqlTimeout;
+
+        private readonly ILog _log;
 
         private int StartingRowOffset { get; set; }
+        private bool _gotTheLastBatch;
+
         private string AssetPairId { get; }
+        private string SearchToken { get; }
+        private DateTime? MigrateByDate { get; }
 
         public TradesSqlHistoryRepository(
             string sqlConnString,
             int sqlQueryBatchSize,
+            TimeSpan sqlTimeout,
             ILog log,
-            int startingRowOffset,
-            string assetPairId
+            DateTime? migrateByDate,
+            string assetPairId,
+            string searchToken
             )
         {
             _sqlQueryBatchSize = sqlQueryBatchSize;
+            _sqlTimeout = sqlTimeout;
             _log = log;
 
-            StartingRowOffset = startingRowOffset;
+            StartingRowOffset = 0; // Will read everything.
+
             AssetPairId = assetPairId;
+            SearchToken = searchToken;
+            MigrateByDate = migrateByDate;
 
             _sqlConnection = new SqlConnection(sqlConnString);
             _sqlConnection.Open();
         }
         
-        public async Task<IEnumerable<TradeHistoryItem>> GetNextBatchAsync()
+        public async Task<IReadOnlyCollection<TradeHistoryItem>> GetNextBatchAsync()
         {
             var result = new List<TradeHistoryItem>();
 
+            // If we got the last batch in the previous iteration, there is no reason to execute one more query
+            // with empty result. Just return.
+            if (_gotTheLastBatch)
```

Comment:
Open connection right before request and then close it. It is no sense to have it always opened.
        _sqlConnection = new SqlConnection(sqlConnString);
             _sqlConnection.Open();	             _sqlConnection.Open();

Reply:
Done.